cmd/singularity/cli/oci.go

```diff
@@ -0,0 +1,748 @@
+				return
+			}
+
+			c.Close()
+
+			switch state.Status {
+			case "created":
+				// ignore error there and wait for stopped status
+				ociStart(containerID)
+			case "running":
+				status <- state.Status
+			case "stopped":
+				status <- state.Status
+			}
+		}
+	}()
+
+	// wait running status
+	s := <-status
+	if s != "running" {
+		return fmt.Errorf("%s", s)
+	}
+
+	engineConfig, err := getEngineConfig(containerID)
+	if err != nil {
+		return err
+	}
+
+	if err := attach(engineConfig, true); err != nil {
+		return err
+	}
+
+	// wait stopped status
+	s = <-status
+	if s != "stopped" {
+		return fmt.Errorf("%s", s)
+	}
+
+	return nil
+}
+
+func ociAttach(containerID string) error {
+	engineConfig, err := getEngineConfig(containerID)
+	if err != nil {
+		return err
+	}
+
+	defer exitContainer(containerID, false)
+
+	return attach(engineConfig, false)
+}
+
+func ociStart(containerID string) error {
+	state, err := getState(containerID)
+	if err != nil {
+		return err
+	}
+
+	if state.Status != "created" {
+		return fmt.Errorf("container %s is not created", containerID)
+	}
+
+	if state.ControlSocket == "" {
+		return fmt.Errorf("can't find control socket")
+	}
+
+	ctrl := &ociruntime.Control{}
+	ctrl.StartContainer = true
+
+	c, err := unix.Dial(state.ControlSocket)
+	if err != nil {
+		return fmt.Errorf("failed to connect to control socket")
+	}
+	defer c.Close()
+
+	enc := json.NewEncoder(c)
+	if err != nil {
+		return err
+	}
+
+	if err := enc.Encode(ctrl); err != nil {
+		return err
+	}
+
+	// wait runtime close socket connection for ACK
+	d := make([]byte, 1)
+	if _, err := c.Read(d); err != io.EOF {
+		return err
+	}
+
+	return nil
+}
+
+func ociKill(containerID string, killTimeout int) error {
+	// send signal to the instance
+	state, err := getState(containerID)
+	if err != nil {
+		return err
+	}
+
+	if state.Status != "created" && state.Status != "running" {
+		return fmt.Errorf("container %s is nor created nor running", containerID)
+	}
+
+	sig := syscall.SIGTERM
+
+	if stopSignal != "" {
+		sig, err = signal.Convert(stopSignal)
+		if err != nil {
+			return err
+		}
+	}
+
+	if killTimeout > 0 {
+		c, err := unix.Dial(state.ControlSocket)
+		if err != nil {
+			return fmt.Errorf("failed to connect to control socket")
+		}
+		defer c.Close()
+
+		killed := make(chan bool, 1)
+
+		go func() {
+			// wait runtime close socket connection for ACK
+			d := make([]byte, 1)
+			if _, err := c.Read(d); err == io.EOF {
+				killed <- true
+			}
+		}()
+
+		if err := syscall.Kill(state.Pid, sig); err != nil {
+			return err
+		}
+
+		select {
+		case <-killed:
+		case <-time.After(time.Duration(killTimeout) * time.Second):
+			return syscall.Kill(state.Pid, syscall.SIGKILL)
+		}
+	} else {
+		return syscall.Kill(state.Pid, sig)
+	}
+
+	return nil
+}
+
+func ociDelete(containerID string) error {
+	engineConfig, err := getEngineConfig(containerID)
+	if err != nil {
+		return err
+	}
+
+	switch engineConfig.State.Status {
+	case "running":
```

Comment:
Should we have all statuses as a constants somewhere? E.g in `pkg/ociruntime` since it is already present

Reply:
Good idea